### PR TITLE
Tracing for eventhubs consumer and batch processor

### LINKED ISSUE #28559: [FEATURE REQ] EventProcessorClient batch receive does not support tracing
We only trace single event processing, there is no tracing for batch case

https://github.com/Azure/azure-sdk-for-java/blob/c39d7d22b7e18a42e6df370957ecc0f14f7d4e05/sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/PartitionPumpManager.java

### PR DESCRIPTION
**Pull request by lmolkova**
Fixes #28559, #28552

Similar to #30508 and enables the same user experience shown on the screenshots there

### CODE REVIEW

File: eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml [hunk 1]
```diff
@@ -100,6 +100,7 @@
   <suppress checks="IllegalImport" files=".*[/\\]com[/\\]azure[/\\]identity[/\\]*"/>
   <suppress checks="IllegalImport" files="com.azure.messaging.servicebus.TracingIntegrationTests.java"/>
   <suppress checks="IllegalImport" files="com.azure.messaging.eventhubs.TracingIntegrationTests.java"/>
+  <suppress checks="IllegalImport" files="com.azure.messaging.eventhubs.PublishEventsTracingWithCustomContextSample.java"/>
```
Comment (conniey) [inline eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml:103]: I am curious, what is the illegal import it doesn't like?

File: sdk/core/azure-core-tracing-opentelemetry/README.md [hunk 2]
```diff
@@ -122,9 +122,13 @@ try {
 
 Send a single event/message using [azure-messaging-eventhubs][azure-messaging-eventhubs] with tracing enabled.
 
-Users can additionally pass the value of the current tracing span to the EventData object with key **PARENT_TRACE_CONTEXT_KEY** on the [Context][context] object:
+Users can additionally pass custom value of the trace context to the EventData object with key **PARENT_TRACE_CONTEXT_KEY** on the [Context][context] object.
+
+Please refer to [EventHubs samples](https://github.com/Azure/azure-sdk-for-java/blob/10a18ccc2f20cad6004ae90d64f22009d65e9ef7/sdk/eventhubs/azure-messaging-eventhubs/src/samples/java/com/azure/messaging/eventhubs/PublishEventsTracingWithCustomContextSample.java)
```
Comment (conniey) [inline sdk/core/azure-core-tracing-opentelemetry/README.md:127]: Will we have to update this to main in the future?

Comment (conniey) [reply sdk/core/azure-core-tracing-opentelemetry/README.md:127]: I'd try to use the link style of the ones below. Also, it is "Event Hubs".

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsConsumerInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        if (tracer.isEnabled() && !isSync) {
+            child = tracer.startSpan(spanName, tracer.setParentAndAttributes(message, enqueuedTime, parent), ProcessKind.PROCESS);
+        }
+
+        meter.reportReceive(enqueuedTime, partitionId, child);
+
+        return child;
+    }
+
+    private Instant getEnqueuedTime(MessageAnnotations messageAnnotations) {
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsConsumerInstrumentation.java:51]: iirc, EventHubMessageSerializer has a method that does this similar logic.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsTracer.java [hunk 4]
```diff
@@ -0,0 +1,255 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsTracer.java:1]: Do we have any tests for this class?

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsConsumerInstrumentation.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (conniey) [inline sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsConsumerInstrumentation.java:1]: Similarly, is it possible to mock some tests for this class?

File: eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml (hunk 1 above)
Comment (lmolkova) [reply eng/code-quality-reports/src/main/resources/checkstyle/checkstyle-suppressions.xml:103]: The sample imports OTel stuff and we disallow it in https://github.com/lmolkova/azure-sdk-for-java/blob/4a77508c86ae1f1698f86cecd65986129892062b/eng/code-quality-reports/src/main/resources/checkstyle/checkstyle.xml#L66-L67

It's a necessary check for library code, but not for samples/tests

File: sdk/core/azure-core-tracing-opentelemetry/README.md (hunk 2 above)
Comment (lmolkova) [reply sdk/core/azure-core-tracing-opentelemetry/README.md:127]: > Will we have to update this to main in the future?

yes, unfortunately, I need to provide a working link first and it doesn't exist in main yet.

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsTracer.java (hunk 4 above)
Comment (lmolkova) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsTracer.java:1]: Yes, it's tested with existing tracing tests (that used to test old helper in core-amqp) plus new unit tests, plus new integration tests. The overall test coverage for this is 93%

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsConsumerInstrumentation.java (hunk 5 above)
Comment (lmolkova) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsConsumerInstrumentation.java:1]: It is tested with metrics and tracing tests, 100% coverage

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsTracer.java (hunk 4 above)
Comment (lmolkova) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsTracer.java:1]: image

File: sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsConsumerInstrumentation.java (hunk 3 above)
Comment (lmolkova) [reply sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/instrumentation/EventHubsConsumerInstrumentation.java]: good catch! I move it to MessageUtils and used in both places.

**Review by conniey — APPROVED** [review]
Thanks!

**Pull request merged**